File: src/harnesses/claudecode/hooks/on-hooks.py
#!/usr/bin/env python3
# =============================================================================
# src/harnesses/claudecode/hooks/on-hooks.py
# Generic claudecode hook dispatcher — reads every module's hooks.json manifest
# and routes claudecode hook input to the declared commands.
#
# Usage (invoked from .claude/settings.local.json):
#   python3 src/harnesses/claudecode/hooks/on-hooks.py <phase>
#
# Phases:
#   pre-tool   → PreToolUse (Bash)          → command.before
#   post-file  → PostToolUse (Edit|Write)   → file.edited
#   post-bash  → PostToolUse (Bash)         → command.after
#   stop       → Stop                       → session.idle
#   startup    → SessionStart               → session.created
# =============================================================================

import datetime
import json
import os
import re
import subprocess
import sys

DEV_BOT_ROOT = os.path.abspath(os.path.join(os.path.dirname(os.path.realpath(__file__)), "..", "..", "..", ".."))


def load_manifests():
    base = os.path.join(DEV_BOT_ROOT, "src", "agentic")
    hooks = []
    if not os.path.isdir(base):
        return hooks
    for d in sorted(os.listdir(base)):
        mf = os.path.join(base, d, "hooks.json")
        if not os.path.isfile(mf):
            continue
        try:
            data = json.load(open(mf))
        except Exception:
            continue
        for hook in data.get("hooks", []):
            hook["_module"] = os.path.join(base, d)
            hooks.append(hook)
    return hooks


def resolve(run, ctx):
    return [re.sub(r"\{(\w[\w-]*)\}", lambda m: ctx.get(m.group(1), ""), p) for p in run]


def quiet(cmd, cwd):
    subprocess.run(cmd, cwd=cwd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)


def deny(reason):
    # Claude Code's PreToolUse hook contract: the legacy top-level "decision"
    # field is not recognized (only approve/block, and the current schema uses
    # hookSpecificOutput.permissionDecision) — a plain {"decision":"deny"} is
    # silently IGNORED and the command runs. Emit the current schema so guards
    # actually block.
    print(json.dumps({
        "hookSpecificOutput": {
            "hookEventName": "PreToolUse",
            "permissionDecision": "deny",
            "permissionDecisionReason": "Command blocked by guard rule: " + reason,
        },
    }))


def run_and_log(cmd, cwd, log_path, hook_id):
    try:
        r = subprocess.run(cmd, cwd=cwd, capture_output=True, text=True)
        text = (r.stdout or "").strip()
        if text:
            os.makedirs(os.path.dirname(log_path), exist_ok=True)
            ts = datetime.datetime.now(datetime.timezone.utc).isoformat()
            with open(log_path, "a") as f:
                f.write(f"[{ts}] {hook_id}\n{text}\n\n")
    except Exception:
        pass


def run_hook(hook, ctx, worktree):
    # Route a hook's output to its declared "log" path when present, else run
    # quietly. Every phase (file.edited, command.after, session.idle,
    # session.created) must honor the manifest's log field — audit-31 §5 found
    # the non-post-file phases discarding output unconditionally, so a
    # session-start prune that declared qmd-index.log never wrote an entry and
    # the self-heal looked broken.
    cmd = resolve(hook["run"], ctx)
    log_path = hook.get("log")
    if log_path:
        run_and_log(cmd, worktree, os.path.join(worktree, log_path), hook["id"])
    else:
        quiet(cmd, worktree)


def run_file_edits(file_path, worktree):
    for hook in load_manifests():
        if hook.get("event") != "file.edited" or not hook.get("run"):
            continue
        match = hook.get("match", {})
        if "file" in match and not re.search(match["file"], file_path):
            continue
        if "content" in match:
            try:
                head = open(file_path).read(4096)
                if not re.search(match["content"], head):
                    continue
            except Exception:
                continue
        run_hook(hook, {"module": hook["_module"], "worktree": worktree, "file": file_path}, worktree)


def main():
    phase = sys.argv[1] if len(sys.argv) > 1 else ""
    data = json.load(sys.stdin)
    worktree = data.get("cwd") or os.getcwd()

    if phase == "pre-tool":
        command = (data.get("tool_input") or {}).get("command") or ""
        tool_name = (data.get("tool_name") or "")
        for hook in load_manifests():
            if hook.get("event") != "command.before" or not hook.get("run"):
                continue
            # Match filtering (same semantics as run_file_edits): a hook scoped
            # to specific tools (e.g. guards' match.tool: ["bash","shell"]) must
            # not fire for every tool. Without this, a future second
            # command.before hook would silently run on every tool call.
            # Case-insensitive: Claude Code reports tool_name as "Bash" while
            # the manifest declares lowercase "bash".
            match = hook.get("match", {})
            if "tool" in match:
                allowed = {t.lower() for t in match["tool"]}
                if tool_name.lower() not in allowed:
                    continue
            ctx = {
                "module": hook["_module"],
                "worktree": worktree,
                "command": command,
                "agent": "",
                "global-config": os.path.join(DEV_BOT_ROOT, ".devbot.global.jsonc"),
                "project-config": os.path.join(worktree, ".devbot.project.jsonc"),
            }
            cmd = resolve(hook["run"], ctx)
            try:
                r = subprocess.run(cmd, cwd=worktree, capture_output=True, text=True)
            except Exception:
                # The guard tool failed to execute (missing interpreter/script,
                # transiently absent during a reinit rebuild). Claude Code's
                # PreToolUse contract treats a hook execution ERROR as "no
                # decision" → default-allow — so a blocking guard that cannot
                # run must fail CLOSED (audit-31 §2), not let the command
                # through.
                if hook.get("blocking"):
                    deny("guard temporarily unavailable (failed to execute)")
                    return
                continue
            if hook.get("blocking"):
                try:
                    parsed = json.loads(r.stdout or "{}")
                    if r.returncode != 0:
                        # Non-zero exit = the guard itself errored; no valid
                        # decision was produced. Fail closed.
                        deny("guard temporarily unavailable (exited %d)" % r.returncode)
                        return
                    if parsed.get("blocked"):
                        deny(parsed.get("message") or "guard rule")
                        return
                except Exception:
                    # Unparseable output from a blocking guard — no decision
                    # was produced. Fail closed rather than silently allow.
                    deny("guard temporarily unavailable (unparseable output)")
                    return

    elif phase == "post-file":
        file_path = (data.get("tool_input") or {}).get("file_path") or ""
        if file_path:
            run_file_edits(file_path, worktree)

    elif phase == "post-bash":
        command = (data.get("tool_input") or {}).get("command") or ""
        exit_code = (data.get("tool_output") or {}).get("exit_code")
        if isinstance(exit_code, (int, float)) and exit_code != 0:
            return
        try:
            hash_ = subprocess.check_output(
                ["git", "-C", worktree, "log", "-1", "--format=%H"],
                text=True, stderr=subprocess.DEVNULL,
            ).strip()
        except Exception:
            hash_ = ""
        for hook in load_manifests():
            if hook.get("event") != "command.after" or not hook.get("run"):
                continue
            match = hook.get("match", {})
            if "command" in match and not re.search(match["command"], command):
                continue
            run_hook(hook, {"module": hook["_module"], "worktree": worktree, "command": command, "hash": hash_}, worktree)

    elif phase == "stop":
        session_id = data.get("session_id") or "default"
        for hook in load_manifests():
            if hook.get("event") != "session.idle" or not hook.get("run"):
                continue
            run_hook(hook, {"module": hook["_module"], "worktree": worktree, "session-id": session_id}, worktree)

    elif phase == "startup":
        for hook in load_manifests():
            if hook.get("event") != "session.created" or not hook.get("run"):
                continue
            run_hook(hook, {"module": hook["_module"], "worktree": worktree}, worktree)


if __name__ == "__main__":
    main()

File: src/harnesses/opencode/hooks/on-hooks.ts
// src/harnesses/opencode/hooks/on-hooks.ts
// Generic harness adapter: reads every module's `hooks.json` manifest and wires
// its declared hooks into opencode's plugin API. Business logic stays in each
// module's `tools/` entry; this adapter maps events → commands (or delegates
// `plugin:` hooks).
//
// IMPORTANT: this module must export ONLY the OnHooks plugin factory. opencode
// invokes every function export of a registered plugin as a plugin factory, so
// any helper export crashes or corrupts the plugin — helpers live in the
// sibling ../on-hooks-utils.ts (outside hooks/, so the harness init never
// registers it as a plugin).

import type { Plugin } from "@opencode-ai/plugin"
import { readFileSync, readdirSync, existsSync } from "fs"
import { join } from "path"
import { execSync } from "child_process"
import { createLogger } from "../../../_shared/logger.ts"
import { defaultHookLog, createFileEditGate, guardDecision, resolveGlobalConfigPath, routeHookOutput, type HookDecl } from "../on-hooks-utils"

const DEV_BOT_ROOT = join(import.meta.dir, "../../../..") // repo root

function loadManifests(): { moduleDir: string; hooks: HookDecl[] }[] {
  const base = join(DEV_BOT_ROOT, "src", "agentic")
  if (!existsSync(base)) return []
  const out: { moduleDir: string; hooks: HookDecl[] }[] = []
  for (const dir of readdirSync(base)) {
    const manifest = join(base, dir, "hooks.json")
    if (!existsSync(manifest)) continue
    try {
      const parsed = JSON.parse(readFileSync(manifest, "utf8"))
      if (Array.isArray(parsed?.hooks)) out.push({ moduleDir: join(base, dir), hooks: parsed.hooks })
    } catch {
      // Skip malformed manifest.
    }
  }
  return out
}

function resolve(run: string[], ctx: Record<string, string>): string[] {
  return run.map((part) => part.replace(/\{(\w[\w-]*)\}/g, (_, key: string) => ctx[key] ?? ""))
}

async function runCommand(cmd: string[], cwd: string): Promise<{ stdout: string; stderr: string; exitCode: number }> {
  const proc = Bun.spawn(cmd, { cwd, stdout: "pipe", stderr: "pipe" })
  // Drain both streams concurrently — an undrained stderr pipe fills up and
  // blocks the child hook once the buffer exceeds ~64KB.
  const [stdout, stderr] = await Promise.all([
    new Response(proc.stdout).text(),
    new Response(proc.stderr).text(),
  ])
  const exitCode = await proc.exited
  return { stdout, stderr, exitCode }
}

// (appendLog, defaultHookLog and routeHookOutput live in ../on-hooks-utils.ts
// — see the header note about why this module cannot export them.)

export const OnHooks: Plugin = async ({ directory, worktree, project, client }) => {
  const root = worktree || project?.worktree || directory
  const logger = createLogger({
    module: "hooks",
    // audit-24 NOTE-2: route logger output to the hooks log file so hook
    // failures are captured even when plugin stderr is not.
    logFile: join(root, defaultHookLog()),
  })
  const manifests = loadManifests()
  // Per-file gate for file.edited hooks (audit-32: burst-edit race).
  const fileEditGate = createFileEditGate()

  // Plugin-type hooks declared in module manifests: import each factory and
  // merge the handlers it returns into this adapter's dispatch. This is how
  // client-using hooks (auto-recover prompt injection, the stall watchdog)
  // stay module-owned while being wired for the opencode harness.
  const pluginHandlers: Record<string, Array<(input: any, output?: any) => Promise<void>>> = {}
  for (const { moduleDir, hooks } of manifests) {
    for (const hook of hooks) {
      if (!hook.plugin) continue
      const pluginPath = resolve([hook.plugin], { module: moduleDir })[0]
      const abs = pluginPath.startsWith("/") ? pluginPath : join(DEV_BOT_ROOT, pluginPath)
      try {
        const mod = await import(abs)
        const factory = mod?.default
        if (typeof factory !== "function") {
          logger.debug(`plugin hook ${hook.id}: no default export at ${abs}`)
          continue
        }
        const handlers = await factory({ directory, worktree, project, client })
        if (!handlers) continue
        for (const [eventName, fn] of Object.entries(handlers)) {
          if (typeof fn !== "function") continue
          ;(pluginHandlers[eventName] ??= []).push(fn as any)
        }
        logger.info(`loaded plugin hook ${hook.id} from ${abs}`)
      } catch (e) {
        logger.debug(`plugin hook ${hook.id} failed to load: ${e instanceof Error ? e.message : String(e)}`)
      }
    }
  }

  // Non-blocking event dispatch (file.edited, session.*).
  async function dispatch(eventName: string, matchFile: string | undefined, ctx: Record<string, string>) {
    for (const { moduleDir, hooks } of manifests) {
      for (const hook of hooks) {
        if (hook.event !== eventName || !hook.run) continue
        const re = hook.match?.file
        if (re && matchFile !== undefined && !new RegExp(re).test(matchFile)) continue
        const contentRe = hook.match?.content
        if (contentRe && matchFile !== undefined) {
          try {
            const head = readFileSync(matchFile, "utf-8").slice(0, 4096)
            if (!new RegExp(contentRe).test(head)) continue
          } catch {
            continue
          }
        }
        const cmd = resolve(hook.run, { module: moduleDir, worktree: root, ...ctx })
        try {
          const out = await runCommand(cmd, root)
          routeHookOutput(out, hook, root)
        } catch (e) {
          logger.debug(`${hook.id} hook failed: ${e instanceof Error ? e.message : String(e)}`)
        }
      }
    }
  }

  return {
    event: async ({ event }: { event: any }) => {
      for (const fn of pluginHandlers.event ?? []) await fn({ event })
      const type = event?.type
      if (type === "file.edited") {
        const file: string = (event as any).properties?.file ?? ""
        if (file) {
          // audit-32 FAIL: format-yml burst race — two file.edited events
          // within ~1s ran two concurrent format hooks whose read-modify-write
          // interleaves corrupted the file. Serialize + coalesce per file: a
          // burst collapses into one in-flight run plus one trailing re-run.
          fileEditGate(file, () => dispatch("file.edited", file, { file }))
        }
      } else if (type === "session.created") {
        await dispatch("session.created", undefined, {})
      } else if (type === "session.idle") {
        const id = event?.properties?.sessionID || event?.properties?.info?.id || event?.properties?.id || ""
        await dispatch("session.idle", undefined, { "session-id": id })
      } else if (type === "session.error") {
        const id = event?.properties?.sessionID || event?.properties?.info?.id || event?.properties?.id || ""
        const err = event?.properties?.error?.message || event?.properties?.error?.data?.message || event?.properties?.message || ""
        await dispatch("session.error", undefined, { "session-id": id, error: err })
      }
    },

    "tool.execute.before": async (input: any, output: any) => {
      for (const fn of pluginHandlers["tool.execute.before"] ?? []) await fn(input, output)
      const tool = String(input?.tool ?? "").toLowerCase()
      const command = String((output?.args as any)?.command ?? (input?.args as any)?.command ?? "")
      for (const { moduleDir, hooks } of manifests) {
        for (const hook of hooks) {
          if (hook.event !== "command.before" || !hook.run) continue
          if (hook.match?.tool && !hook.match.tool.includes(tool)) continue

          const agent = process.env.OPENCODE_AGENT ?? ""
          // audit-32 FAIL: guards silently disabled — env is DEV_BOT_ROOT, not
          // DEVBOT_ROOT. Resolve from the plugin's own root (realpath, already
          // used for manifest loading above) so guards run with rules.
          const globalConfig = resolveGlobalConfigPath(DEV_BOT_ROOT, process.env)
          const projectConfig = join(root, ".devbot.project.jsonc")
          const cmd = resolve(hook.run, {
            module: moduleDir,
            worktree: root,
            command,
            agent,
            "global-config": globalConfig,
            "project-config": projectConfig,
          })
          try {
            const out = await runCommand(cmd, root)
            const decision = guardDecision(out, hook.blocking)
            if (decision.blocked) {
              throw new Error(`[${hook.id}] Command blocked: ${decision.message}`)
            }
          } catch (e) {
            if (e instanceof Error && e.message.startsWith(`[${hook.id}]`)) throw e
            // The guard tool failed to execute (missing interpreter/script).
            // A blocking guard that cannot run must fail CLOSED, not silently
            // let the command through (audit-31 §2) — opencode treats an
            // exception from tool.execute.before as a block.
            if (hook.blocking) {
              throw new Error(`[${hook.id}] Command blocked: guard temporarily unavailable (failed to execute)`)
            }
            logger.debug(`${hook.id} hook failed: ${e instanceof Error ? e.message : String(e)}`)
          }
        }
      }
    },

    "tool.execute.after": async (input: any, output: any) => {
      for (const fn of pluginHandlers["tool.execute.after"] ?? []) await fn(input, output)
      const tool = String(input?.tool ?? "").toLowerCase()
      if (tool !== "bash" && tool !== "shell") return
      const command = String(input?.args?.command ?? "")

      // Fire only on successful commands.
      const ec = output?.metadata?.exitCode
      if (typeof ec === "number" && ec !== 0) return
      const outStr = String(output?.output ?? "")
      if (/fatal:/i.test(outStr) || /Aborting/i.test(outStr)) return

      let hash = ""
      try {
        hash = execSync('git log -1 --format="%H"', { cwd: root, encoding: "utf8", timeout: 5000 }).trim()
      } catch {
        // No commit hash available.
      }

      for (const { moduleDir, hooks } of manifests) {
        for (const hook of hooks) {
          if (hook.event !== "command.after" || !hook.run) continue
          if (hook.match?.command && !new RegExp(hook.match.command).test(command)) continue
          const cmd = resolve(hook.run, { module: moduleDir, worktree: root, command, hash })
          try {
            const out = await runCommand(cmd, root)
            // Route output like dispatch() does — a command.after hook that
            // declares a "log" field must not have its output silently
            // discarded (mirrors the claudecode post-bash fix, audit-31 §5).
            routeHookOutput(out, hook, root)
          } catch (e) {
            logger.debug(`${hook.id} hook failed: ${e instanceof Error ? e.message : String(e)}`)
          }
        }
      }
    },
  }
}
